remote-keypair/src/openpgp_card.rs

```diff
@@ -0,0 +1,359 @@
+/// card supporting OpenPGP that it finds connected to the machine.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Locator {
+    pub aid: Option<ApplicationIdentifier>,
+}
+
+#[derive(Clone, Debug, Error, PartialEq, Eq)]
+pub enum LocatorError {
+    #[error("failed to parse OpenPGP AID: {0}")]
+    IdentifierParseError(String),
+    #[error(transparent)]
+    UriReferenceError(#[from] URIReferenceError),
+    #[error("mismatched scheme")]
+    MismatchedScheme,
+}
+
+impl TryFrom<&URIReference<'_>> for Locator {
+    type Error = LocatorError;
+
+    /// Extract a locator from a URI.
+    ///
+    /// "pgpcard://" => Default locator.
+    ///
+    /// "pgpcard://D2760001240103040006123456780000" => Locator pointing to a
+    /// OpenPGP instance identifiable by AID D2760001240103040006123456780000.
+    /// As per the OpenPGP specification:
+    ///   * AID must be a 16-byte hexadecimal string (32 digits).
+    ///   * D276000124 => AID preamble, fixed across all AIDs
+    ///   * 01         => application ID, fixed to 01 == OpenPGP
+    ///   * 0304       => OpenPGP version, 3.4 in this case
+    ///   * 0006       => unique manufacturer ID, Yubico in this case
+    ///   * 12345678   => smart card serial number
+    ///   * 0000       => reserved for future use
+    ///
+    /// No other URI formats are valid.
+    fn try_from(uri: &URIReference<'_>) -> Result<Self, Self::Error> {
+        let scheme = uri.scheme().map(|s| s.as_str().to_ascii_lowercase());
+        let ident = uri.host().map(|h| h.to_string());
+
+        match (scheme, ident) {
+            (Some(scheme), Some(ident)) if scheme == "pgpcard" => {
+                if ident.is_empty() {
+                    return Ok(Self { aid: None });
+                }
+                if ident.len() != 32 {
+                    return Err(LocatorError::IdentifierParseError(
+                        "OpenPGP AID must be 32 digits".to_string(),
+                    ));
+                }
+                let mut ident_bytes = Vec::<u8>::new();
+                for hex_byte in &ident.chars().chunks(2) {
+                    ident_bytes.push(
+                        u8::from_str_radix(hex_byte.collect::<String>().as_str(), 16).map_err(
+                            |_| {
+                                LocatorError::IdentifierParseError(
+                                    "non-hex character found in identifier".to_string(),
+                                )
+                            },
+                        )?,
+                    );
+                }
+                Ok(Self {
+                    aid: Some(ident_bytes.as_slice().try_into().map_err(|_| {
+                        LocatorError::IdentifierParseError("invalid identifier format".to_string())
+                    })?),
+                })
+            }
+            (Some(scheme), None) if scheme == "pgpcard" => Ok(Self { aid: None }),
+            _ => Err(LocatorError::MismatchedScheme),
+        }
```

Comment:
It seems that `scheme` has to be `gpgcard`.  In which case, removing it from the `match` could
reduce the indentation and mental load of the reader:

```rust
        if !matches!(scheme, Some("pgpcard")) {
            return Err(LocatorError::MismatchedScheme);
        }

        let ident = match ident {
            None => return Ok(Self { aid: None }),
            Some(ident) if ident.is_empty() => return Ok(Self { aid: None }),
            Some(ident) => ident,
        };
        
        if ident.len() != 32 {
            return Err(LocatorError::IdentifierParseError(
                "OpenPGP AID must contain 32 hexadecimal characters".to_string(),
            ));
        }

        let ident_bytes = ident
            .chars()
            .chunks(2)
            .into_iter()
            .map(|hex_byte| {
                u8::from_str_radix(hex_byte.collect::<String>().as_str(), 16).map_err(
                    |_| {
                        LocatorError::IdentifierParseError(
                            "non-hex character found in the URL path".to_string(),
                        )
                    },
                )
            })
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self {
            aid: Some(ident_bytes.as_slice().try_into().map_err(|_| {
                LocatorError::IdentifierParseError("invalid OpenPGP URL format".to_string())
            })?),
        })
```

Should the `MismatchedScheme` error include the actual scheme name for clarity?
I also wonder if it would make sense to have a separate error for the case when the scheme is
absent.

```rust
        match scheme {
            None => return Err(LocatorError::SchemeRequired),
            Some(scheme) if scheme != "pgpcard" => {
                return Err(LocatorError::MismatchedScheme(scheme.to_owned()));
            }
            _ => (),
        }
```

Similarly, it might make sense to include failed hex character and the failed URL component in the
parsing errors.